In [1]:
print("hello world")

hello world


In [2]:
import pandas as pd

df = pd.read_csv('train.csv')
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


In [9]:
mask = df["Age"] > 70
df[mask]

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
96,97,0,1,"Goldschmidt, Mr. George B",male,71.0,0,0,PC 17754,34.6542,A5,C
116,117,0,3,"Connors, Mr. Patrick",male,70.5,0,0,370369,7.7500,NaN,Q
493,494,0,1,"Artagaveytia, Mr. Ramon",male,71.0,0,0,PC 17609,49.5042,NaN,C
630,631,1,1,"Barkworth, Mr. Algernon Henry Wilson",male,80.0,0,0,27042,30.0000,A23,S
851,852,0,3,"Svensson, Mr. Johan",male,74.0,0,0,347060,7.7750,NaN,S


In [ ]:
df[mask]['Age'] #아래거랑 똑같은 의미

96     71.0
116    70.5
493    71.0
630    80.0
851    74.0
Name: Age, dtype: float64

In [12]:
df.loc[mask, ['Age', 'Name']]

,Age,Name
96,71.0,"Goldschmidt, Mr. George B"
116,70.5,"Connors, Mr. Patrick"
493,71.0,"Artagaveytia, Mr. Ramon"
630,80.0,"Barkworth, Mr. Algernon Henry Wilson"
851,74.0,"Svensson, Mr. Johan"


In [13]:
import pandas as pd
data = {'이름': ['철수', '영희', '민수', '수지', '호영'],
        '나이': [25, 32, 40, 28, 18],
        '도시': ['서울', '대전', '인천', '대구', '부산']}
df = pd.DataFrame(data)
print(df)

   이름  나이  도시
0  철수  25  서울
1  영희  32  대전
2  민수  40  인천
3  수지  28  대구
4  호영  18  부산


In [18]:
mask = df['나이'] >= 30
df.loc[mask]

,이름,나이,도시
1,영희,32,대전
2,민수,40,인천


In [20]:
mask2 = (df['도시'] == '서울') & (df['나이'] >= 20)
df.loc[mask2]

,이름,나이,도시
0,철수,25,서울


In [3]:
# Titanic 데이터 로드
import pandas as pd
titanic_df = pd.read_csv('train.csv')
titanic_df.head(5)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


In [24]:
# 예제 1
mask = titanic_df['Age'] > 70
titanic_df.loc[mask, ['Name', 'Age', 'Survived']]

,Name,Age,Survived
96,"Goldschmidt, Mr. George B",71.0,0
116,"Connors, Mr. Patrick",70.5,0
493,"Artagaveytia, Mr. Ramon",71.0,0
630,"Barkworth, Mr. Algernon Henry Wilson",80.0,1
851,"Svensson, Mr. Johan",74.0,0


In [4]:
# 예제 2
conditions = (titanic_df['Pclass'] == 1) & (titanic_df['Sex'] == 'female')
titanic_df.loc[conditions, ['Name','Pclass','Sex']]

,Name,Pclass,Sex
1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",1,female
3,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",1,female
11,"Bonnell, Miss. Elizabeth",1,female
31,"Spencer, Mrs. William Augustus (Marie Eugenie)",1,female
52,"Harper, Mrs. Henry Sleeper (Myna Haxtun)",1,female
...,...,...,...
856,"Wick, Mrs. George Dennick (Mary Hitchcock)",1,female
862,"Swift, Mrs. Frederick Joel (Margaret Welles Ba...",1,female
871,"Beckwith, Mrs. Richard Leonard (Sallie Monypeny)",1,female
879,"Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)",1,female


In [5]:
titanic_df['Pclass'].value_counts()

Pclass
3    491
1    216
2    184
Name: count, dtype: int64

In [28]:
titanic_df['Sex'].value_counts()

Sex
male      577
female    314
Name: count, dtype: int64

In [32]:
# 예제 3
mask1 = (titanic_df['Age'] < 15) | (titanic_df['Age'] >= 80)
titanic_df[mask1]

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
7,8,0,3,"Palsson, Master. Gosta Leonard",male,2.00,3,1,349909,21.0750,NaN,S
9,10,1,2,"Nasser, Mrs. Nicholas (Adele Achem)",female,14.00,1,0,237736,30.0708,NaN,C
10,11,1,3,"Sandstrom, Miss. Marguerite Rut",female,4.00,1,1,PP 9549,16.7000,G6,S
14,15,0,3,"Vestrom, Miss. Hulda Amanda Adolfina",female,14.00,0,0,350406,7.8542,NaN,S
16,17,0,3,"Rice, Master. Eugene",male,2.00,4,1,382652,29.1250,NaN,Q
...,...,...,...,...,...,...,...,...,...,...,...,...
827,828,1,2,"Mallet, Master. Andre",male,1.00,0,2,S.C./PARIS 2079,37.0042,NaN,C
831,832,1,2,"Richards, Master. George Sibley",male,0.83,1,1,29106,18.7500,NaN,S
850,851,0,3,"Andersson, Master. Sigvard Harald Elias",male,4.00,4,2,347082,31.2750,NaN,S
852,853,0,3,"Boulos, Miss. Nourelain",female,9.00,1,1,2678,15.2458,NaN,C


In [34]:
titanic_df[mask1].describe()

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,79.000000,79.000000,79.000000,79.000000,79.000000,79.000000,79.000000
mean,426.341772,0.582278,2.607595,6.755316,1.797468,1.303797,31.682859
std,281.904097,0.496335,0.628750,9.385692,1.651571,0.667072,27.178358
min,8.000000,0.000000,1.000000,0.420000,0.000000,0.000000,7.229200
25%,172.500000,0.000000,2.000000,2.000000,0.000000,1.000000,18.750000
50%,436.000000,1.000000,3.000000,4.000000,1.000000,1.000000,27.750000
75%,689.500000,1.000000,3.000000,9.000000,3.500000,2.000000,32.193750
max,870.000000,1.000000,3.000000,80.000000,5.000000,2.000000,151.550000


In [ ]:
age_20_30 = titanic_df[titanic_df["Age"].between(20, 30)]
# 첫값 끝값 포함

In [ ]:
temp_isin = titanic_df[titanic_df["Embarked"].isin(["C","Q"])]
# | 사용하는거랑 비슷

In [ ]:
temp_na = titanic_df[(titanic_df["Age"] < 16) | (titanic_df["Age"].isna())]
# 조건에 결측인 행까지 포함할지 말지

In [ ]:
# 체인 인덱싱
titanic_df[titanic_df["Age"] > 70]["Fare"] = 0
# 원본이 바뀔 수도 복사본만 바뀔 수도 있음

C:\Users\j2h09\AppData\Local\Temp\ipykernel_25848\420371582.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  titanic_df[titanic_df["Age"] > 70]["Fare"] = 0


In [36]:
titanic_df.loc[titanic_df["Age"] > 70, 'Fare'] = 0
titanic_df.loc[titanic_df["Age"] > 70, 'Fare']

96     0.0
116    0.0
493    0.0
630    0.0
851    0.0
Name: Fare, dtype: float64

In [ ]:
titanic_df.sort_values(by="Fare", ascending=False)
# 내림차순

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
258,259,1,1,"Ward, Miss. Anna",female,35.0,0,0,PC 17755,512.3292,NaN,C
737,738,1,1,"Lesurer, Mr. Gustave J",male,35.0,0,0,PC 17755,512.3292,B101,C
679,680,1,1,"Cardeza, Mr. Thomas Drake Martinez",male,36.0,0,1,PC 17755,512.3292,B51 B53 B55,C
438,439,0,1,"Fortune, Mr. Mark",male,64.0,1,4,19950,263.0000,C23 C25 C27,S
341,342,1,1,"Fortune, Miss. Alice Elizabeth",female,24.0,3,2,19950,263.0000,C23 C25 C27,S
...,...,...,...,...,...,...,...,...,...,...,...,...
116,117,0,3,"Connors, Mr. Patrick",male,70.5,0,0,370369,0.0000,NaN,Q
822,823,0,1,"Reuchlin, Jonkheer. John George",male,38.0,0,0,19972,0.0000,NaN,S
815,816,0,1,"Fry, Mr. Richard",male,NaN,0,0,112058,0.0000,B102,S
806,807,0,1,"Andrews, Mr. Thomas Jr",male,39.0,0,0,112050,0.0000,A36,S


In [ ]:
# - 열 삭제 - `df.drop(columns='열')`(거의 안 쓰긴 함)
# - 행 삭제 - `df.drop(index=인덱스번호)` 
# - 열 이름 변경 - `df.rename(columns={'현재 열': '바꿀 열'})`
# 원본 바꾸고 싶다면 원본 df에 받아줘야 함

In [38]:
data2 = {'이름': ['철수', '영희', '민수', '수지'],
         '점수': [85, 92, 78, 92]}
df2 = pd.DataFrame(data2)
df2

,이름,점수
0,철수,85
1,영희,92
2,민수,78
3,수지,92


In [39]:
# 점수 정렬 내림차순
df2.sort_values(by='점수', ascending=False)

,이름,점수
1,영희,92
3,수지,92
0,철수,85
2,민수,78


In [41]:
# 점수 정렬 내림차순, 이름 정렬 오름차순
df2.sort_values(by=['점수', '이름'], ascending= [False, True])

,이름,점수
3,수지,92
1,영희,92
0,철수,85
2,민수,78


In [42]:
df2['메모'] = ['임시', '필요없음', '임시', '필요없음']  # 예시로 메모 열 추가
print(df2)

   이름  점수    메모
0  철수  85    임시
1  영희  92  필요없음
2  민수  78    임시
3  수지  92  필요없음


In [ ]:
# 메모컬럼 삭제
df2_cleaned = df2.drop(columns = '메모')

# 두번째 행 삭제
df2_cleaned = df2.drop(index = 2)

# 점수라는 컬럼을 새로운 점수라는 컬럼명으로 변경
df2_cleaned = df2.rename(columns={'점수':'새로운 점수'})

,이름,새로운 점수,메모
0,철수,85,임시
1,영희,92,필요없음
2,민수,78,임시
3,수지,92,필요없음


In [47]:
# Titanic 데이터 로드
titanic_df = pd.read_csv('train.csv')
titanic_df.head(5)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


In [51]:
titanic_df.sort_values(by ='Fare', ascending = False)
titanic_df.sort_values(by = ['Age', 'Name'], ascending = [True, True])
titanic_df.drop(columns=['Cabin', 'Ticket'])

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Fare,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,7.2500,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,71.2833,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,7.9250,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,53.1000,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,8.0500,S
...,...,...,...,...,...,...,...,...,...,...
886,887,0,2,"Montvila, Rev. Juozas",male,27.0,0,0,13.0000,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.0,0,0,30.0000,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,NaN,1,2,23.4500,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.0,0,0,30.0000,C


In [52]:
# 결측치가 포함된 예제 Series
s = pd.Series([10, None, 30, None, 50])
print("원본 Series:")
print(s)

원본 Series:
0    10.0
1     NaN
2    30.0
3     NaN
4    50.0
dtype: float64


In [54]:
s.isna().sum()

np.int64(2)

In [55]:
titanic_df.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [59]:
missing_cnt = titanic_df.isna().sum().sort_values(ascending=False)
missing_cnt

Cabin          687
Age            177
Embarked         2
PassengerId      0
Name             0
Pclass           0
Survived         0
Sex              0
Parch            0
SibSp            0
Fare             0
Ticket           0
dtype: int64

In [ ]:
missing_ratio = (titanic_df.isna().mean() * 100).sort_values(ascending=False)
missing_ratio
# 각 컬럼 단위로 계산

Cabin          77.104377
Age            19.865320
Embarked        0.224467
PassengerId     0.000000
Name            0.000000
Pclass          0.000000
Survived        0.000000
Sex             0.000000
Parch           0.000000
SibSp           0.000000
Fare            0.000000
Ticket          0.000000
dtype: float64

In [58]:
missing_cnt, missing_ratio

(Cabin          687
 Age            177
 Embarked         2
 PassengerId      0
 Name             0
 Pclass           0
 Survived         0
 Sex              0
 Parch            0
 SibSp            0
 Fare             0
 Ticket           0
 dtype: int64,
 Cabin          77.104377
 Age            19.865320
 Embarked        0.224467
 PassengerId     0.000000
 Name            0.000000
 Pclass          0.000000
 Survived        0.000000
 Sex             0.000000
 Parch           0.000000
 SibSp           0.000000
 Fare            0.000000
 Ticket          0.000000
 dtype: float64)

In [61]:
# 결측치가 포함된 예제 Series
s = pd.Series([10, None, 30, None, 50])
print("원본 Series:")
print(s)

원본 Series:
0    10.0
1     NaN
2    30.0
3     NaN
4    50.0
dtype: float64


In [62]:
s.dropna()

0    10.0
2    30.0
4    50.0
dtype: float64

In [63]:
titanic_df.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [64]:
s.fillna('unknown')

0       10.0
1    unknown
2       30.0
3    unknown
4       50.0
dtype: object

In [65]:
s.fillna(s.median())

0    10.0
1    30.0
2    30.0
3    30.0
4    50.0
dtype: float64

In [69]:
titanic_df['Age'].fillna(titanic_df['Age'].median())

0      22.0
1      38.0
2      26.0
3      35.0
4      35.0
       ... 
886    27.0
887    19.0
888    28.0
889    26.0
890    32.0
Name: Age, Length: 891, dtype: float64

In [71]:
mask = titanic_df['Embarked'].isna()
titanic_df.loc[mask]

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
61,62,1,1,"Icard, Miss. Amelie",female,38.0,0,0,113572,80.0,B28,NaN
829,830,1,1,"Stone, Mrs. George Nelson (Martha Evelyn)",female,62.0,0,0,113572,80.0,B28,NaN


- **Embarked (2건 결측):** 어떻게 처리할래?
    - 삭제
    - 최빈값으로 채운다.
- **Age (177건 결측):** 어떻게 처리할래?
    - 그룹별 중앙값, 평균으로 대체
- **Cabin (687건 결측):** 어떻게 처리할래?
    - 컬럼 자체 드롭하거나 Unknown으로 대체

In [72]:
df3 = pd.DataFrame({'제품': ['A', 'B', 'B', 'C'],
                    '가격': [1000, 2000, 2000, 3000]})
print(df3)

  제품    가격
0  A  1000
1  B  2000
2  B  2000
3  C  3000


In [74]:
df3.duplicated(keep='first').sum()

np.int64(1)

In [75]:
cleaned = df3.duplicated()
cleaned

0    False
1    False
2     True
3    False
dtype: bool

In [76]:
import pandas as pd
duplicated_titanic_df = pd.read_csv("titanic_with_duplicates.csv")
duplicated_titanic_df.head()

,PassengerId,survived,pclass,sex,age,sibsp,parch,fare,embarked
0,8,1,3,female,30.0,0,0,13.00,S
1,1,0,3,male,22.0,1,0,7.25,S
2,5,0,3,male,28.0,0,0,8.05,S
3,1,0,3,male,22.0,1,0,7.25,S
4,6,1,2,female,27.0,0,0,10.50,S


In [77]:
duplicated_titanic_df.shape

(13, 9)

In [79]:
duplicated_titanic_df.duplicated().sum()

np.int64(5)

In [80]:
duplicated_titanic_df.drop_duplicates()

,PassengerId,survived,pclass,sex,age,sibsp,parch,fare,embarked
0,8,1,3,female,30.0,0,0,13.0000,S
1,1,0,3,male,22.0,1,0,7.2500,S
2,5,0,3,male,28.0,0,0,8.0500,S
4,6,1,2,female,27.0,0,0,10.5000,S
6,2,1,1,female,38.0,1,0,71.2833,C
8,4,0,1,male,35.0,1,0,53.1000,S
10,7,0,3,male,19.0,0,0,7.8958,S
11,3,1,3,female,26.0,0,0,7.9250,S


In [81]:
import pandas as pd
import numpy as np
dup_df = pd.DataFrame({
    "PassengerId": [1, 1, 2, 2, 2, 3, 3],
    "Name":        ["Allen, Mr.", "Allen, Mr.", "Baker, Mr.", "Baker, Mr.", "Baker, Mr.", "Clark, Mrs.", "Clark, Mrs."],
    "Age":         [22, 22, np.nan, 35, 35, 29, np.nan],
    "Fare":        [7.25, 7.25, 8.05, np.nan, 8.05, 71.28, 71.28],
    "Embarked":    ["S", "S", "S", "S", "S", "C", "C"],
})
dup_df

,PassengerId,Name,Age,Fare,Embarked
0,1,"Allen, Mr.",22.0,7.25,S
1,1,"Allen, Mr.",22.0,7.25,S
2,2,"Baker, Mr.",NaN,8.05,S
3,2,"Baker, Mr.",35.0,NaN,S
4,2,"Baker, Mr.",35.0,8.05,S
5,3,"Clark, Mrs.",29.0,71.28,C
6,3,"Clark, Mrs.",NaN,71.28,C


In [82]:
temp = dup_df.copy()

In [84]:
temp.isna().sum(axis=1)

0    0
1    0
2    1
3    1
4    0
5    0
6    1
dtype: int64

In [86]:
temp['na_count'] = temp.isna().sum(axis=1)
temp

,PassengerId,Name,Age,Fare,Embarked,na_count
0,1,"Allen, Mr.",22.0,7.25,S,0
1,1,"Allen, Mr.",22.0,7.25,S,0
2,2,"Baker, Mr.",NaN,8.05,S,1
3,2,"Baker, Mr.",35.0,NaN,S,1
4,2,"Baker, Mr.",35.0,8.05,S,0
5,3,"Clark, Mrs.",29.0,71.28,C,0
6,3,"Clark, Mrs.",NaN,71.28,C,1


In [88]:
best_rows = (
    temp.sort_values(["PassengerId", "na_count"])   # 결측 적은 행이 위로
       .drop_duplicates(subset=["PassengerId"], keep="first")  # PassengerId별 1개만 남김
       .drop(columns=["na_count"])
)

best_rows

,PassengerId,Name,Age,Fare,Embarked
0,1,"Allen, Mr.",22.0,7.25,S
4,2,"Baker, Mr.",35.0,8.05,S
5,3,"Clark, Mrs.",29.0,71.28,C
